### PR TITLE
Sync user federation

### PR DESCRIPTION
**Pull request by JaquierM**
Pull request for issue #438

### CODE REVIEW

**Comment by jkroepke** [comment]
@JaquierM whats next here?

PR is still in draft, jobs failing.

**Comment by JaquierM** [comment]
Something came up at work and didn't leave me time to continue last week. I will finish this PR this week.

**Review by jkroepke — COMMENTED** [review]
Hi,

thanks for providing the PR.

Please provide a test where `import.sync-user-federation=true` is set, but no userstorageproviders are configured. That should increase the coverage.

Should the sync run on providers even if import Users is set to false?

I have a question. Did you think is its necessary to setup a whole openldap server? Or is it good enough to check for a specific error message? (I thinking about test execution time and required resource on the system for an additional container)

**Comment by JaquierM** [comment]
Hi,

> Should the sync run on providers even if import Users is set to false?

I'm not sure about what parameter you talk about. There is no import.users or import.managed.user parameters.
Can you clarify it for me please?

**Comment by jkroepke** [comment]
> I'm not sure about what parameter you talk about.

this one.

image

**Comment by JaquierM** [comment]
> Please provide a test where import.sync-user-federation=true is set, but no userstorageproviders are configured. That should increase the coverage.

Done

> Should the sync run on providers even if import Users is set to false?

Indeed, it is better to not run it to avoid unnecessary request.

> I have a question. Did you think is its necessary to setup a whole openldap server? Or is it good enough to check for a specific error message? (I thinking about test execution time and required resource on the system for an additional container)

It is not necessary. The test is now base on the error 500 return by Keycloak.

File: src/test/java/de/adorsys/keycloak/config/service/SyncUserFederationIT.java [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+        } catch (Throwable t) {
+            throwable = t;
+        }
+
+        assertThat(throwable, new IsNot<>(new IsNull<>()));
+        // This matching use the name of the function where the error occurs in order to guarantee the source.
+        boolean throwableCameFromUserFederationSync = Arrays.stream(throwable.getStackTrace())
+                .anyMatch(stackTraceElement -> stackTraceElement.getMethodName().contains(FUNCTION_WHERE_THE_ERROR_MUST_BE_GENERATED));
+        assertThat(throwable, new ThrowableMessageMatcher<>(is("HTTP 500 Internal Server Error")));
+        assertThat(throwableCameFromUserFederationSync, is(true));
```
Comment (jkroepke) [inline src/test/java/de/adorsys/keycloak/config/service/SyncUserFederationIT.java:66]: One question here: 

Why assertThrows wont fit here? Like https://github.com/adorsys/keycloak-config-cli/blob/e9cefdcfba9a38aaf4e38e1ef15f399b7bca04d9/src/test/java/de/adorsys/keycloak/config/service/ImportSimpleRealmIT.java#L196-L202

**Review by jkroepke — COMMENTED** [review]
Looks good to me in general 👍 

But I want to know, if there is a reason why assertThrow wouldn't fit here.

File: src/test/java/de/adorsys/keycloak/config/service/SyncUserFederationIT.java (hunk 1 above)
Comment (JaquierM) [reply src/test/java/de/adorsys/keycloak/config/service/SyncUserFederationIT.java:66]: It will fit perfectly. I will make the modification today.

**Review by jkroepke — APPROVED** [review]
Thanks!

**Pull request merged**